File: solutions/0203/1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (curr != nullptr) {
            ListNode* next = curr->next;
            if (curr->val == val) {
                if (prev == nullptr) {
                    head = next;
                } else {
                    prev->next = next;
                }
                curr = next;
            } else {
                prev = curr;
                curr = next;
            }
        }

        return head;
    }
};

File: solutions/0082/1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            if (curr.next != null && curr.val == curr.next.val) {
                ListNode q = curr;
                while (q != null && q.val == curr.val) {
                    q = q.next;
                }
                if (prev == null) {
                    head = q;
                } else {
                    prev.next = q;
                }
                curr = q;
            } else {
                prev = curr;
                curr = curr.next;
            }
        }
        return head;
    }
}
